src/main/kotlin/fr/xebia/xke/5_FP.kt

```diff
@@ -0,0 +1,71 @@
+package fr.xebia.xke
+
+import arrow.core.*
+import arrow.syntax.option.toEither
+
+import arrow.core.Either
+import arrow.core.Option
+import arrow.core.flatMap
+import arrow.data.NonEmptyList
+import arrow.data.Validated
+import arrow.data.applicative
+import arrow.data.ev
+
+/**
+ * This class validates user data in two ways:
+ * - Applicative for all fields (validated all fields)
+ * - Monadic within each field (stops on the first error per field)
+ */
+class UserValidation {
+
+    /*
+     * The following functions might help you
+     * - Option.fromNullable("something")
+     * - Option(maybeRight).toEither { "Left value" }
+     * - Either.cond(true, { 42 }, { "Error" })
+     * - Validated.fromEither(...)
+     * - Validated.applicative<List<String>>()
+            .map2(validated1, validated2, { v: Tuple<Int, Int> -> v.a + v.b })
+            .ev()
+     * - NonEmptyList.of("a string", "another string")
+     */
+
+    // TODO find a way to aggregate in an applicative way the 'readName' and 'readAge' functions
+    fun validateUser(params: Map<String, String>): Validated<NonEmptyList<String>, User>  {
+        return Validated.applicative<NonEmptyList<String>>()
+            .map2(
+                Validated.fromEither(readName(params)),
+                Validated.fromEither(readAge(params)),
+                { z: Tuple2<String, Int> -> User(z.a, z.b) }
+            ).ev()
+    }
+
+    // TODO read the field 'name' and validate that is non blank
+    fun readName(params: Map<String, String>): Either<NonEmptyList<String>, String> =
+        getValue("name", params)
+            .flatMap { nonBlank ("name", it) }
+
+    // TODO read the field 'age' and validate that is non blank, valid int, and non negative
+    fun readAge(params: Map<String, String>): Either<NonEmptyList<String>, Int> =
+        getValue("age", params)
+            .flatMap { nonBlank ("age", it) }
+            .flatMap { parseInt ("age", it) }
+            .flatMap { nonNegative (it) }
+
+    // TODO get a value from the map or return an error
+    fun getValue(name: String, data: Map<String, String>): Either<NonEmptyList<String>, String> =
+        Option.fromNullable(data[name]).toEither { NonEmptyList.of("$name is not present") }
```

Comment:
I prefer `data[name]?.let { Either.right(it) } ?: Either.left("$name is not present")`

Reply:
Ok, why not